testplan/common/utils/monitor/collectors.py

```diff
@@ -0,0 +1,141 @@
+"""Collect system information"""
+
+import os
+import time
+import platform
+import subprocess
+import multiprocessing
+import psutil
+from testplan.common.utils.path import pwd
+
+
+class Collector(object):
+    def __init__(self, director):
+        self.metadata = {
+            'num_cpu': multiprocessing.cpu_count(),
+            'memory': psutil.virtual_memory().total,
+            'disk_total': psutil.disk_usage(director).total,
+            'disk_director': director,
+            'system': platform.platform(),
+            'extra': {},
+        }
+        self._pid = os.getpid()
+
+        # save last io counter per process
+        self._io_counter = {}
+
+    def monitor(self):
+        parent = psutil.Process(self._pid)
+        procs = parent.children(recursive=True)
+        procs.append(parent)
+        monitor_result = {
+            'cpu': 0,
+            'memory': 0,
+            'disk': psutil.disk_usage(self.metadata['disk_director']).percent,
+            'iops': 0,
+            'read': 0,
+            'write': 0,
+        }
+        for proc in procs:
+            try:
+                raw_data = proc.as_dict(
+                    attrs=[
+                        'pid',
+                        'name',
+                        'memory_info',
+                        'cpu_percent',
+                        'io_counters',
+                        'create_time',
+                        'io_counters'
+                    ])
+
+            except psutil.NoSuchProcess:
+                continue
+            monitor_result['cpu'] += raw_data['cpu_percent'] \
+                if raw_data['cpu_percent'] > 0 else 0
```

Comment:
CPU percent can be negative...?

Reply:
It's a known issue of psutil on Linux

Reply:
You mean https://github.com/giampaolo/psutil/issues/392 ? Looks like it was fixed. It's fine to be defensive just in case though, was mostly just curious.